### PR TITLE
Support for VMware Cloud Director as a cloud provider

### LINKED ISSUE #9924: Support for VMware Cloud Director in KKP
- [x] Support to create, delete new clusters
- [x] Add support in API

Additional endpoints will be covered with:
https://github.com/kubermatic/kubermatic/issues/9939

### PR DESCRIPTION
**Pull request by ahmedwaleedmalik**
Signed-off-by: Waleed Malik <ahmedwaleedmalik@gmail.com>

**What does this PR do / Why do we need it**:
Support for VMware Cloud Director as a cloud provider. This covers the required parts for KKP and the API.

**Does this PR close any issues?**:
Fixes #9924

**Special notes for your reviewer**:
Additional endpoints will be added in a follow up PR against https://github.com/kubermatic/kubermatic/issues/9939

**Documentation**:

**Does this PR introduce a user-facing change?**:

```release-note
Support for VMware Cloud Director as a cloud provider
Bump machine-controller to v1.51.0
```

### CODE REVIEW

**Comment by ahmedwaleedmalik** [comment]
Blocked by https://github.com/kubernetes-sigs/aws-iam-authenticator/pull/460

**Comment by embik** [comment]
> Blocked by [kubernetes-sigs/aws-iam-authenticator#460](https://github.com/kubernetes-sigs/aws-iam-authenticator/pull/460)

Also see https://github.com/kubermatic/kubermatic/pull/9826. I think we were planning to remove aws-iam-authenticator from our dependencies in that PR.

File: go.mod [hunk 1]
```diff
@@ -58,64 +58,69 @@ require (
...
+	kubevirt.io/containerized-data-importer-api v1.49.0
 	sigs.k8s.io/aws-iam-authenticator v0.5.8
-	sigs.k8s.io/controller-runtime v0.11.2
+	sigs.k8s.io/controller-runtime v0.12.1
 	sigs.k8s.io/controller-tools v0.8.0
 	sigs.k8s.io/yaml v1.3.0
 )
 
+// Added for testing purposes will be removed before merging this PR
+replace sigs.k8s.io/aws-iam-authenticator => github.com/ahmedwaleedmalik/aws-iam-authenticator v0.5.8-0.20220531123014-13bd59d5c7e9
```
Comment (ahmedwaleedmalik) [inline go.mod]: This will be removed before merging this PR. In favor of either https://github.com/kubermatic/kubermatic/pull/9826 or https://github.com/kubernetes-sigs/aws-iam-authenticator/pull/460, whichever gets merged first.

File: go.mod [hunk 2]
```diff
@@ -113,9 +114,12 @@ replace (
 	k8s.io/cluster-bootstrap => k8s.io/cluster-bootstrap v0.23.6
 	k8s.io/code-generator => k8s.io/code-generator v0.23.6
 	k8s.io/kube-aggregator => k8s.io/kube-aggregator v0.23.6
+	k8s.io/kube-openapi => k8s.io/kube-openapi v0.0.0-20220124234850-424119656bbf
 	k8s.io/kubectl => k8s.io/kubectl v0.23.6
 	k8s.io/kubelet => k8s.io/kubelet v0.23.6
 	k8s.io/metrics => k8s.io/metrics v0.23.6
+	k8s.io/utils => k8s.io/utils v0.0.0-20211116205334-6203023598ed
+	sigs.k8s.io/controller-runtime => sigs.k8s.io/controller-runtime v0.11.2
```
Comment (xrstf) [inline go.mod:122]: Why update this to 0.12 and then override it with 0.11.2 again?

File: pkg/controller/master-controller-manager/rbac/test/fake/informer_factory.go [hunk 3]
```diff
@@ -29,6 +29,10 @@ type SharedInformerFactory struct {
 	kubeinformers.SharedInformerFactory
 }
 
+func (i *dummySharedIndexInformer) SetTransform(handler cache.TransformFunc) error {
+	panic("implement me")
+}
+
```
Comment (xrstf) [inline pkg/controller/master-controller-manager/rbac/test/fake/informer_factory.go]: Isn't this only if we had 1.24 dependencies?

File: pkg/provider/cloud/vmwareclouddirector/provider.go [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+	}
+
+	org, err := client.GetOrganization()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get organization %s: %w", client.Auth.Organization, err)
+	}
+
+	vdc, err := client.GetVDCForOrg(*org)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Organization VDC '%s': %w", client.Auth.VDC, err)
```
Comment (xrstf) [inline pkg/provider/cloud/vmwareclouddirector/provider.go]: settle on whether to capitalize "Organization" or not ;)

File: pkg/provider/cloud/vmwareclouddirector/provider.go [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+	// Ensure that Organization exists.
+	org, err := client.GetOrganization()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get organization %s: %w", client.Auth.Organization, err)
+	}
+
+	// Ensure that VDC exists.
+	vdc, err := client.GetVDCForOrg(*org)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Organization VDC '%s': %w", client.Auth.VDC, err)
```
Comment (xrstf) [inline pkg/provider/cloud/vmwareclouddirector/provider.go]: capitalization

File: pkg/provider/cloud/vmwareclouddirector/provider.go [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+	// We need to validate VDC as well since we scope everything down to VDC.
+	// It's not consumed in the credentials directly.
+	org, err := client.GetOrganization()
+	if err != nil {
+		return fmt.Errorf("failed to get organization %s: %w", client.Auth.Organization, err)
+	}
+
+	_, err = client.GetVDCForOrg(*org)
+	if err != nil {
+		return fmt.Errorf("failed to get Organization VDC '%s': %w", vdc, err)
```
Comment (xrstf) [inline pkg/provider/cloud/vmwareclouddirector/provider.go]: capitalization

File: pkg/resources/prometheus/configmap-rules.go [hunk 7]
```diff
@@ -274,8 +274,8 @@ groups:
     labels:
       kubermatic: federate
 
-  - record: job:machine_controller_machines:sum
-    expr: sum(machine_controller_machines)
+  - record: job:machine_controller_machines_total:rate5m
+    expr: rate(machine_controller_machines_total[5m])
```
Comment (xrstf) [inline pkg/resources/prometheus/configmap-rules.go:278]: Why did you change this?

Comment (ahmedwaleedmalik) [reply pkg/resources/prometheus/configmap-rules.go:278]: The metric name was changed from `machine_controller_machines` to `machine_controller_machines_total`

File: go.mod (hunk 2 above)
Comment (ahmedwaleedmalik) [reply go.mod:122]: We have to pin it because it gets updated to `0.12` because of https://github.com/kubermatic/machine-controller/blob/v1.51.0/go.mod#L56

File: pkg/resources/prometheus/configmap-rules.go (hunk 7 above)
Comment (ahmedwaleedmalik) [reply pkg/resources/prometheus/configmap-rules.go:278]: https://github.com/kubermatic/machine-controller/blob/master/pkg/controller/machine/metrics.go#L104
